### ISSUE TITLE
User is unable to create namespaces if tenant ownership is directly assigned

### ISSUE THREAD

**Issue by micke-post** [clastix/capsule#708]
# Bug description

We set up capsule on an AKS cluster configured to use Azure AD authentication with Azure RBAC. On this cluster we set up a few tenants with test users. One of the users is owner of two tenants - one directly, and one via inheritance from an AAD group. The user is able to create namespaces in the tenant that is owned by a group they are a member of, but the one they are a direct owner of.

# How to reproduce
Capsule and the groups are set up like this:

```
> az ad group show --group 93d72ae7-2de0-4b3c-b446-d4cc0f424e03 | ConvertFrom-Json | fl id, displayname
id          : 93d72ae7-2de0-4b3c-b446-d4cc0f424e03
displayName : apptest_k8s_capsule1

> az ad group member list --group 93d72ae7-2de0-4b3c-b446-d4cc0f424e03  | ConvertFrom-Json | fl displayname, id
displayName : TESTUSER K8sTestUser1
id          : 41dddc31-2b85-47d5-a4ce-b82ffdc9a090

displayName : TESTUSER K8sTestUser3
id          : 0fb550a3-cc65-49d8-b9da-e6ae457ea6ba

displayName : TESTUSER K8sTestUser2
id          : 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0
```

```
> helm get values capsule -n capsule-system
USER-SUPPLIED VALUES:
manager:
  options:
    capsuleUserGroups:
    - 93d72ae7-2de0-4b3c-b446-d4cc0f424e03
    forceTenantPrefix: true
```	

The tenants are set up like this:
```
> kubectl get tenants
NAME                  STATE    NAMESPACE QUOTA   NAMESPACE COUNT   NODE SELECTOR   AGE
testtenant1           Active                     1                                 5d
testtenant2           Active                     0                                 24h

> kubectl get tenant testtenant1 -o yaml --insecure-skip-tls-verify
apiVersion: capsule.clastix.io/v1beta2
kind: Tenant
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"capsule.clastix.io/v1beta2","kind":"Tenant","metadata":{"annotations":{},"name":"testtenant1"},"spec":{"owners":[{"kind":"Group","name":"93d72ae7-2de0-4b3c-b446-d4cc0f424e03"}]}}
  creationTimestamp: "2023-02-10T10:13:14Z"
  generation: 2
  name: testtenant1
  resourceVersion: "4632018"
  uid: 886ca9cc-c0d8-4a9f-a3fb-0144c7cb87a2
spec:
  owners:
  - clusterRoles:
    - admin
    - capsule-namespace-deleter
    kind: Group
    name: 93d72ae7-2de0-4b3c-b446-d4cc0f424e03
status:
  namespaces:
  - testtenant1-test1
  - testtenant1-test2
  - testtenant1-test3
  size: 3
  state: Active

> kubectl get tenant testtenant2 -o yaml --insecure-skip-tls-verify
apiVersion: capsule.clastix.io/v1beta2
kind: Tenant
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"capsule.clastix.io/v1beta2","kind":"Tenant","metadata":{"annotations":{},"name":"testtenant2"},"spec":{"owners":[{"kind":"User","name":"991d45f2-b5c7-4b83-b570-f4fa87c4a6b0"}]}}
  creationTimestamp: "2023-02-14T09:30:12Z"
  generation: 1
  name: testtenant2
  resourceVersion: "4030693"
  uid: ec984205-05c6-4635-a814-58f7bd343448
spec:
  owners:
  - clusterRoles:
    - admin
    - capsule-namespace-deleter
    kind: User
    name: 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0
status:
  size: 0
  state: Active
```

If I now log in with user 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0, I expect to be able to create namespaces in both tenants
```
> az ad signed-in-user show | ConvertFrom-Json | fl id
id : 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0

> kubectl config set-cluster coaks --server=https://coaks.westeurope.cloudapp.azure.com:443
Cluster "coaks" set.
> kubectl config set-context coaks --cluster=coaks --user=clusterUser_aks-westeurope_aks-westeurope-cluster
Context "coaks" modified.
> kubectl config use-context coaks
Switched to context "coaks".

> kubectl create ns testtenant1-test3 --insecure-skip-tls-verify
namespace/testtenant1-test3 created

> kubectl create ns testtenant2-test3 --insecure-skip-tls-verify
Error from server (The namespace doesn't match the tenant prefix, expected testtenant1-testtenant2-test3): admission webhook "namespaces.capsule.clastix.io" denied the request: The namespace doesn't match the tenant prefix, expected testtenant1-testtenant2-test3
```
I also tried creating the following file
```
kind: Namespace
apiVersion: v1
metadata:
  name: testtenant2-d1
  labels:
    capsule.clastix.io/tenant: testtenant2
```
And deploying it with kubectl
```
> kubectl apply -f <file name>
Error from server (Cannot assign the desired namespace to a non-owned Tenant): error when creating "<file name>": admission webhook "owner.namespace.capsule.clastix.io" denied the request: Cannot assign the desired namespace to a non-owned Tenant
```

# Expected behavior
The user should be able to create namespaces in all owned tenants

# Logs
Below are the logs from running kubectl create ns testtenant1-test3  and kubectl create ns testtenant2-test3 
```
{"level":"debug","ts":"2023-02-15T14:00:52.694Z","logger":"controller-runtime.webhook.webhooks","msg":"received request","webhook":"/nodes","UID":"d822100e-9480-4b0e-8f77-7bbe64696f8b","kind":"/v1, Kind=Node","resource":{"group":"","version":"v1","resource":"nodes"}}
{"level":"debug","ts":"2023-02-15T14:00:52.694Z","logger":"controller-runtime.webhook.webhooks","msg":"wrote response","webhook":"/nodes","code":200,"reason":"","UID":"d822100e-9480-4b0e-8f77-7bbe64696f8b","allowed":true}
{"level":"debug","ts":"2023-02-15T14:00:56.484Z","logger":"controller-runtime.webhook.webhooks","msg":"received request","webhook":"/namespace-owner-reference","UID":"675843d8-5db4-4f8d-bf56-1fa2d6d42f99","kind":"/v1, Kind=Namespace","resource":{"group":"","version":"v1","resource":"namespaces"}}
{"level":"debug","ts":"2023-02-15T14:00:56.484Z","logger":"controller-runtime.webhook.webhooks","msg":"wrote response","webhook":"/namespace-owner-reference","code":200,"reason":"","UID":"675843d8-5db4-4f8d-bf56-1fa2d6d42f99","allowed":true}
{"level":"debug","ts":"2023-02-15T14:00:56.484Z","logger":"events","msg":"Normal","object":{"kind":"Tenant","name":"testtenant1","uid":"886ca9cc-c0d8-4a9f-a3fb-0144c7cb87a2","apiVersion":"capsule.clastix.io/v1beta2","resourceVersion":"4563574"},"reason":"NamespaceCreationWebhook","message":"Namespace testtenant1-test3 has been assigned to the desired Tenant"}
{"level":"debug","ts":"2023-02-15T14:00:56.505Z","logger":"controller-runtime.webhook.webhooks","msg":"received request","webhook":"/namespaces","UID":"157330a6-1fb7-4676-bb66-cbe45cdb1176","kind":"/v1, Kind=Namespace","resource":{"group":"","version":"v1","resource":"namespaces"}}
{"level":"debug","ts":"2023-02-15T14:00:56.505Z","logger":"controller-runtime.webhook.webhooks","msg":"wrote response","webhook":"/namespaces","code":200,"reason":"","UID":"157330a6-1fb7-4676-bb66-cbe45cdb1176","allowed":true}
{"level":"info","ts":"2023-02-15T14:00:56.525Z","logger":"controllers.Tenant","msg":"Ensuring limit resources count is updated","Request.Name":"testtenant1"}
{"level":"info","ts":"2023-02-15T14:00:56.525Z","logger":"controllers.Tenant","msg":"Ensuring all Namespaces are collected","Request.Name":"testtenant1"}
{"level":"info","ts":"2023-02-15T14:00:56.534Z","logger":"controllers.Tenant","msg":"Starting processing of Namespaces","Request.Name":"testtenant1","items":3}
{"level":"debug","ts":"2023-02-15T14:00:56.535Z","logger":"events","msg":"Normal","object":{"kind":"Tenant","name":"testtenant1","uid":"886ca9cc-c0d8-4a9f-a3fb-0144c7cb87a2","apiVersion":"capsule.clastix.io/v1beta2","resourceVersion":"4632018"},"reason":"testtenant1-test1","message":"Ensuring Namespace metadata"}
{"level":"debug","ts":"2023-02-15T14:00:56.535Z","logger":"events","msg":"Normal","object":{"kind":"Tenant","name":"testtenant1","uid":"886ca9cc-c0d8-4a9f-a3fb-0144c7cb87a2","apiVersion":"capsule.clastix.io/v1beta2","resourceVersion":"4632018"},"reason":"testtenant1-test2","message":"Ensuring Namespace metadata"}
[...]
{"level":"debug","ts":"2023-02-15T14:01:12.448Z","logger":"events","msg":"Normal","object":{"kind":"Tenant","name":"testtenant1","uid":"886ca9cc-c0d8-4a9f-a3fb-0144c7cb87a2","apiVersion":"capsule.clastix.io/v1beta2","resourceVersion":"4632018"},"reason":"NamespaceCreationWebhook","message":"Namespace testtenant2-test3 has been assigned to the desired Tenant"}
{"level":"debug","ts":"2023-02-15T14:01:12.452Z","logger":"controller-runtime.webhook.webhooks","msg":"received request","webhook":"/namespaces","UID":"fa1155ca-9793-40e2-9819-f544930a90d3","kind":"/v1, Kind=Namespace","resource":{"group":"","version":"v1","resource":"namespaces"}}
{"level":"debug","ts":"2023-02-15T14:01:12.452Z","logger":"controller-runtime.webhook.webhooks","msg":"wrote response","webhook":"/namespaces","code":403,"reason":"The namespace doesn't match the tenant prefix, expected testtenant1-testtenant2-test3","UID":"fa1155ca-9793-40e2-9819-f544930a90d3","allowed":false}
{"level":"debug","ts":"2023-02-15T14:01:12.452Z","logger":"events","msg":"Warning","object":{"kind":"Tenant","name":"testtenant1","uid":"886ca9cc-c0d8-4a9f-a3fb-0144c7cb87a2","apiVersion":"capsule.clastix.io/v1beta2","resourceVersion":"4632018"},"reason":"InvalidTenantPrefix","message":"Namespace testtenant2-test3 does not match the expected prefix for the current Tenant"}
{"level":"debug","ts":"2023-02-15T14:01:23.237Z","logger":"controller-runtime.webhook.webhooks","msg":"received request","webhook":"/nodes","UID":"5dc1a4b0-2201-4857-b8e8-5a1fa31395e7","kind":"/v1, Kind=Node","resource":{"group":"","version":"v1","resource":"nodes"}}
{"level":"debug","ts":"2023-02-15T14:01:23.237Z","logger":"controller-runtime.webhook.webhooks","msg":"wrote response","webhook":"/nodes","code":200,"reason":"","UID":"5dc1a4b0-2201-4857-b8e8-5a1fa31395e7","allowed":true}

```

# Additional context
- Helm Chart version:
capsule chart: capsule-0.3.0, app version: 0.2.0
capsule proxy: capsule-proxy-0.3.5, app version: 0.3.2
- Kubernetes version: 
Client Version: v1.26.1
Kustomize Version: v4.5.7

**Comment by prometherion** [comment]
@micke-post, thanks for opening this bug report!

Unfortunately, I couldn't reproduce it in my local development environment, although I didn't use Azure AD as the identity provider.

My Tenant definitions:

```yaml
apiVersion: v1
items:
- apiVersion: capsule.clastix.io/v1beta2
  kind: Tenant
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"capsule.clastix.io/v1beta2","kind":"Tenant","metadata":{"annotations":{},"name":"testtenant1"},"spec":{"owners":[{"kind":"Group","name":"93d72ae7-2de0-4b3c-b446-d4cc0f424e03"}]}}
    creationTimestamp: "2023-02-16T10:25:59Z"
    generation: 1
    name: testtenant1
    resourceVersion: "245607"
    uid: 92efe3c5-82d7-467f-92ee-d5b1f0be8b8c
  spec:
    owners:
    - clusterRoles:
      - admin
      - capsule-namespace-deleter
      kind: Group
      name: 93d72ae7-2de0-4b3c-b446-d4cc0f424e03
  status:
    namespaces:
    - testtenant1-test1
    - testtenant1-test2
    - testtenant1-test3
    size: 3
    state: Active
- apiVersion: capsule.clastix.io/v1beta2
  kind: Tenant
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"capsule.clastix.io/v1beta2","kind":"Tenant","metadata":{"annotations":{},"name":"testtenant2"},"spec":{"owners":[{"kind":"User","name":"991d45f2-b5c7-4b83-b570-f4fa87c4a6b0"}]}}
    creationTimestamp: "2023-02-16T10:26:25Z"
    generation: 1
    name: testtenant2
    resourceVersion: "246051"
    uid: 9d72762b-aec9-4758-ad18-5a398c61a924
  spec:
    owners:
    - clusterRoles:
      - admin
      - capsule-namespace-deleter
      kind: User
      name: 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0
  status:
    namespaces:
    - testtenant2-test3
    size: 1
    state: Active
kind: List
metadata:
  resourceVersion: ""
```

I'm a bit concerned by the following error:
> The namespace doesn't match the tenant prefix, expected testtenant1-testtenant2-test3

Are you sure user `991d45f2-b5c7-4b83-b570-f4fa87c4a6b0` has a Tenant assigned?
What happens if you reissue again the Namespace creation command, prior to deleting the Tenant `testtenant1`?

According to Capsule, there's just a single Tenant eligible to be used by that user.

**Comment by micke-post** [comment]
An error on my end was actually the first thing I considered, but everything looks fine, assignment and permission wise. To verify I just ran this in my current session:

```
> kubectl get tenant testtenant2 -o yaml --insecure-skip-tls-verify
apiVersion: capsule.clastix.io/v1beta2
kind: Tenant
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"capsule.clastix.io/v1beta2","kind":"Tenant","metadata":{"annotations":{},"name":"testtenant2"},"spec":{"owners":[{"kind":"User","name":"991d45f2-b5c7-4b83-b570-f4fa87c4a6b0"}]}}
  creationTimestamp: "2023-02-14T09:30:12Z"
  generation: 1
  name: testtenant2
  resourceVersion: "4030693"
  uid: ec984205-05c6-4635-a814-58f7bd343448
spec:
  owners:
  - clusterRoles:
    - admin
    - capsule-namespace-deleter
    kind: User
    name: 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0
status:
  size: 0
  state: Active
  
  > kubectl get tenant
NAME                  STATE    NAMESPACE QUOTA   NAMESPACE COUNT   NODE SELECTOR   AGE
testtenant2           Active                     0                                 2d5h
```

Account 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0 is the only admin (and user) in testtenant 2. I now also deleted testtenant1 *using an admin account( and logged then back into 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0 to retry the action:

```
> kubectl create ns testtenant2-test7 --insecure-skip-tls-verify
Error from server (You do not have any Tenant assigned: please, reach out to the system administrators): admission webhook "owner.namespace.capsule.clastix.io" denied the request: You do not have any Tenant assigned: please, reach out to the system administrato
```

**Comment by prometherion** [comment]
I tried to replicate your environment using the [CoAKS baseline architecture repo](https://github.com/clastix/coaks-baseline-architecture) with the following groups.

- **myCoAKSAdminGroup**: `690b3e0e-c8f8-415e-978f-b0f1bad75d11`
- **myCoAKSCapsuleGroup**: `357faaa7-71b1-4858-8408-d87d775d4f0b`
- **myCoAKSSolarGroup**: `5e7189f0-5e89-4088-9c5e-cc255119bf53`

> The group `myCoAKSCapsuleGroup` has `myCoAKSSolarGroup` as a member

These are, instead, the users:
- **alice@clastix.io**: `1c958938-c76a-408b-b867-f0e139ab0d94`

These are the following Tenants:

```yaml
apiVersion: v1
items:
- apiVersion: capsule.clastix.io/v1beta2
  kind: Tenant
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"capsule.clastix.io/v1beta2","kind":"Tenant","metadata":{"annotations":{},"name":"testtenant1"},"spec":{"owners":[{"kind":"Group","name":"357faaa7-71b1-4858-8408-d87d775d4f0b"}]}}
    creationTimestamp: "2023-02-17T09:24:20Z"
    generation: 1
    name: testtenant1
    resourceVersion: "3625"
    uid: ede8de0c-edf2-4680-bfe4-2a97f58b849c
  spec:
    owners:
    - clusterRoles:
      - admin
      - capsule-namespace-deleter
      kind: Group
      name: 357faaa7-71b1-4858-8408-d87d775d4f0b
  status:
    namespaces:
    - testtenant1-test1
    - testtenant1-test2
    - testtenant1-test3
    size: 3
    state: Active
- apiVersion: capsule.clastix.io/v1beta2
  kind: Tenant
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"capsule.clastix.io/v1beta2","kind":"Tenant","metadata":{"annotations":{},"name":"testtenant2"},"spec":{"owners":[{"kind":"User","name":"1c958938-c76a-408b-b867-f0e139ab0d94"}]}}
    creationTimestamp: "2023-02-17T09:29:01Z"
    generation: 1
    name: testtenant2
    resourceVersion: "4080"
    uid: e7a44ec7-4c02-4c4f-882b-e9efaa2f4af2
  spec:
    owners:
    - clusterRoles:
      - admin
      - capsule-namespace-deleter
      kind: User
      name: 1c958938-c76a-408b-b867-f0e139ab0d94
  status:
    namespaces:
    - testtenant2-test3
    size: 1
    state: Active
kind: List
metadata:
  resourceVersion: ""
```

The `CapsuleConfiguration` specification:

```yaml
apiVersion: capsule.clastix.io/v1beta2
kind: CapsuleConfiguration
metadata:
  annotations:
    meta.helm.sh/release-name: capsule
    meta.helm.sh/release-namespace: capsule-system
  creationTimestamp: "2023-02-17T09:23:23Z"
  generation: 3
  labels:
    app.kubernetes.io/instance: capsule
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: capsule
    app.kubernetes.io/version: 0.2.1
    helm.sh/chart: capsule-0.3.5
  name: default
  resourceVersion: "3485"
  uid: 09668935-557e-40d5-923c-ec7d58f9c350
spec:
  enableTLSReconciler: true
  forceTenantPrefix: true
  overrides:
    TLSSecretName: capsule-tls
    mutatingWebhookConfigurationName: capsule-mutating-webhook-configuration
    validatingWebhookConfigurationName: capsule-validating-webhook-configuration
  protectedNamespaceRegex: ""
  userGroups:
  - 357faaa7-71b1-4858-8408-d87d775d4f0b
```

Helm releases:

```
NAME            NAMESPACE       REVISION        UPDATED                                 STATUS          CHART                 APP VERSION
capsule         capsule-system  1               2023-02-17 10:23:20.988152759 +0100 CET deployed        capsule-0.3.5         0.2.1      
capsule-proxy   capsule-system  1               2023-02-17 10:28:02.849984178 +0100 CET deployed        capsule-proxy-0.3.6   0.3.3
```

I wasn't able to reproduce your issue, although I have Capsule Proxy and Capsule skewed versions: I tried to switch to your, however, it worked the same way.

Unfortunately, if I'm not able to reproduce it, I cannot help in tackling this. To provide further support I need further information, such as the way the AKS cluster and the AD memberships have been configured, the order in which the Tenants have been created, as well as the Tenants.

> Keep in mind that I noticed, time by time, that a Capsule Proxy installation "breaks" a bit the Capsule Operator that must be restarted. Still investigating this and I'll open an issue to track it down, however, please give a try restart the Capsule operator to check if this solves your issue.

Looking forward to more detailed info to support you! 🤗

**Comment by micke-post** [comment]
To start troubleshooting from a blank slate I deleted and redeployed the entire kubernetes cluster. After that I ran the following script to install capsule and a test namespace:

```
$aksStage = 'test'
$aksSubscription = [Removed]
$aksRg = 'aks-westeurope'
$aksName = 'aks-westeurope-cluster'
$podNamespace = 'capsule-system'

az aks get-credentials --name $aksName --resource-group $aksRg --overwrite-existing
kubectl get namespace

# capsule manager
kubectl create namespace  $podNamespace
helm upgrade --install capsule clastix/capsule `
  --namespace $podNamespace `
  --set manager.options.forceTenantPrefix=true

helm upgrade --install capsule clastix/capsule --namespace $podNamespace --set manager.options.forceTenantPrefix=true --set "manager.options.capsuleUserGroups[0]=f35e0548-8df4-43ad-b5f6-40ef51db1f5c" --set "manager.options.capsuleUserGroups[1]=eef9859e-10a5-411e-9a6d-a00ad740716a" --set "manager.options.capsuleUserGroups[2]=e54b7d0d-19d4-46b8-873d-37e758be794c" --set "manager.options.capsuleUserGroups[3]=85d20d71-46de-4ab9-819a-8e9450fd02d2" --set "manager.options.capsuleUserGroups[4]=93d72ae7-2de0-4b3c-b446-d4cc0f424e03"
Start-Sleep -Seconds 10

# capsule proxy
helm upgrade --install capsule-proxy clastix/capsule-proxy `
  --namespace $podNamespace `
  --set service.type=LoadBalancer `
  --set service.port=443 `
  --set options.oidcUsernameClaim=unique_name 

Start-Sleep -Seconds 10
kubectl --namespace $podNamespace annotate `
  service capsule-proxy service.beta.kubernetes.io/azure-dns-label-name=coaks

kubectl apply -f "$PSScriptRoot/testtenant/testtenant2.txt"

$clusterId=$(az aks show `
  --resource-group $aksRg `
  --name $aksName `
  --query id -o tsv)

$testgroupObjectId=$(az ad group show `
  --group "apptest_k8s_capsule1" `
  --query id `
  --output tsv)

az role assignment create `
  --assignee $testgroupObjectId `
  --role "Azure Kubernetes Service Cluster User Role" `
  --scope $clusterId
```

Testtenant2.txt contains the following:
```
apiVersion: capsule.clastix.io/v1beta2
kind: Tenant
metadata:
  name: testtenant2
spec:
  owners:
  - name: 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0
    kind: User
```

Here's the output of that script:
```
> .\deployment.ps1
The behavior of this command has been altered by the following extension: aks-preview
Merged "aks-westeurope-cluster" as current context in [removed]
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code [removed] to authenticate.
NAME                STATUS   AGE
calico-system       Active   22m
default             Active   23m
gatekeeper-system   Active   22m
kube-node-lease     Active   23m
kube-public         Active   23m
kube-system         Active   23m
tigera-operator     Active   22m
namespace/capsule-system created
Release "capsule" does not exist. Installing it now.
NAME: capsule
LAST DEPLOYED: Mon Feb 20 14:54:06 2023
NAMESPACE: capsule-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
- Capsule Operator Helm Chart deployed:

   # Check the capsule logs
   $ kubectl logs -f deployment/capsule-controller-manager -c manager -n capsule-system


   # Check the capsule logs
   $ kubectl logs -f deployment/capsule-controller-manager -c manager -n capsule-system

- Manage this chart:

   # Upgrade Capsule
   $ helm upgrade capsule -f <values.yaml> capsule -n capsule-system

   # Show this status again
   $ helm status capsule -n capsule-system

   # Uninstall Capsule
   $ helm uninstall capsule -n capsule-system
Release "capsule" has been upgraded. Happy Helming!
NAME: capsule
LAST DEPLOYED: Mon Feb 20 14:54:39 2023
NAMESPACE: capsule-system
STATUS: deployed
REVISION: 2
TEST SUITE: None
NOTES:
- Capsule Operator Helm Chart deployed:

   # Check the capsule logs
   $ kubectl logs -f deployment/capsule-controller-manager -c manager -n capsule-system


   # Check the capsule logs
   $ kubectl logs -f deployment/capsule-controller-manager -c manager -n capsule-system

- Manage this chart:

   # Upgrade Capsule
   $ helm upgrade capsule -f <values.yaml> capsule -n capsule-system

   # Show this status again
   $ helm status capsule -n capsule-system

   # Uninstall Capsule
   $ helm uninstall capsule -n capsule-system
Release "capsule-proxy" does not exist. Installing it now.
NAME: capsule-proxy
LAST DEPLOYED: Mon Feb 20 14:54:55 2023
NAMESPACE: capsule-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
- Capsule-proxy Helm Chart deployed:

   # Check the capsule-proxy logs
   $ kubectl logs -f deployment/capsule-proxy -n capsule-system

- Manage this chart:

   # Upgrade capsule-proxy
   $ helm upgrade capsule-proxy -f <values.yaml> capsule-proxy -n capsule-system

   # Show this status again
   $ helm status capsule-proxy -n capsule-system

   # Uninstall capsule-proxy
   $ helm uninstall capsule-proxy -n capsule-system
service/capsule-proxy annotated
tenant.capsule.clastix.io/testtenant2 created
{
  "canDelegate": null,
  "condition": null,
  "conditionVersion": null,
  "description": null,
  "id": "/subscriptions/[removed]/resourcegroups/aks-westeurope/providers/Microsoft.ContainerService/managedClusters/aks-westeurope-cluster/providers/Microsoft.Authorization/roleAssignments/30d15dfe-5c5a-4572-a0dd-2c6fd344b205",
  "name": "30d15dfe-5c5a-4572-a0dd-2c6fd344b205",
  "principalId": "93d72ae7-2de0-4b3c-b446-d4cc0f424e03",
  "principalType": "Group",
  "resourceGroup": "aks-westeurope",
  "roleDefinitionId": "/subscriptions/[removed]/providers/Microsoft.Authorization/roleDefinitions/4abbcc35-e782-43d8-92c5-2d3f1bd2253f",
  "scope": "/subscriptions/[removed]/resourcegroups/aks-westeurope/providers/Microsoft.ContainerService/managedClusters/aks-westeurope-cluster",
  "type": "Microsoft.Authorization/roleAssignments"
}
```



So basically I installed capsule and the proxy, added some user groups, created a tenant with my test user as the tenant, and gave the user that the group is in the 'Azure Kubernetes Service Cluster User Role' on the cluster. I then logged in to the user with id 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0 which is set as the owner of the tenant and tried to create a namespace:

```
> az ad signed-in-user show | ConvertFrom-Json | select id

id
--
991d45f2-b5c7-4b83-b570-f4fa87c4a6b0

> az aks get-credentials --name $aksName --resource-group $aksRg --overwrite-existing
The behavior of this command has been altered by the following extension: aks-preview

> kubectl config set-cluster coaks --server=https://coaks.westeurope.cloudapp.azure.com:443
Cluster "coaks" set.

> kubectl config set-context coaks --cluster=coaks --user=clusterUser_aks-westeurope_aks-westeurope-cluster
Context "coaks" created.

> kubectl config use-context coaks
Switched to context "coaks".

> kubectl create ns test1 --insecure-skip-tls-verify
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code [removed] to authenticate.

Error from server (You do not have any Tenant assigned: please, reach out to the system administrators): admission webhook "owner.namespace.capsule.clastix.io" denied the request: You do not have any Tenant assigned: please, reach out to the system administrators
```

The group mentioned above that this test user is part of is a regular azure AD security group, without anything that's really noteworthy about it. The users are assigned directly to the group:

```
> az ad group show --group 93d72ae7-2de0-4b3c-b446-d4cc0f424e03 | ConvertFrom-Json | fl id, displayname
id          : 93d72ae7-2de0-4b3c-b446-d4cc0f424e03
displayName : apptest_k8s_capsule1

> az ad group member list --group 93d72ae7-2de0-4b3c-b446-d4cc0f424e03  | ConvertFrom-Json | fl displayname, id
displayName : TESTUSER K8sTestUser1
id          : 41dddc31-2b85-47d5-a4ce-b82ffdc9a090

displayName : TESTUSER K8sTestUser3
id          : 0fb550a3-cc65-49d8-b9da-e6ae457ea6ba

displayName : TESTUSER K8sTestUser2
id          : 991d45f2-b5c7-4b83-b570-f4fa87c4a6b0
```

Anything in particular that you are looking for with the cluster configuration? It's an AKS cluster running kubernetes version 1.24.9, configured with Azure AD authentication with Azure RBAC and local accounts disabled. We are using azure CNI networking with the calico network policy.

**Comment by prometherion** [comment]
@micke-post may I ask you to test the Namespace creation without the Capsule Proxy?

That would be helpful to understand what's the "broken" component in this chain.

**Comment by micke-post** [comment]
Slightly related - but what would be the preferred procedure to disable the proxy for individual users or globally on an aks cluster? I just did the following:

```
> helm uninstall capsule-proxy --namespace $podNamespace
release "capsule-proxy" uninstalled

> kubectl config get-contexts
CURRENT   NAME                     CLUSTER                  AUTHINFO                                            NAMESPACE
*         aks-westeurope-cluster   aks-westeurope-cluster   clusterUser_aks-westeurope_aks-westeurope-cluster
> kubectl config use-context aks-westeurope-cluster
Switched to context "aks-westeurope-cluster".

> kubectl create ns adminnamespace
Error from server (You do not have any Tenant assigned: please, reach out to the system administrators): admission webhook "owner.namespace.capsule.clastix.io" denied the request: You do not have any Tenant assigned: please, reach out to the system administrators
```

The error message is correct in this case, I ran this as an admin user that's not part of any capsule tenant - but I was surprised that I even need one. My expectation was that I would be able to just create namespaces as usual with a cluster admin account after I uninstall the proxy.

**Comment by prometherion** [comment]
I could be wrong, IIRC the admin group which is owning the administrator is declared also in the `CapsuleConfiguration`, and must obey the multi-tenancy rules.

@micke-post I'm having really big trouble troubleshooting your problem because I'm not able to replicate it properly, I think there's something we're missing that is ruining the whole investigation. May I ask you to get in touch privately (LinkedIn, Twitter) to schedule a private session together?

**Comment by micke-post** [comment]
@prometherion Awesome, thanks a lot! I just sent a request on linkedin.

**Comment by micke-post** [comment]
After a call we figured out what the issue was. Basically, there was a mismatch between how we configured our tenant permissions and how aks resolves users. aks expected to have the owning user of the tenant provided using its principal name / unique user name while we configured the tenant with the object ID of the user.
After we changed the problematic tenant from
```
apiVersion: capsule.clastix.io/v1beta2
kind: Tenant
metadata:
  name: testtenant2
spec:
  owners:
  - name: [objectId]
    kind: User
```

to
```
apiVersion: capsule.clastix.io/v1beta2
kind: Tenant
metadata:
  name: testtenant2
spec:
  owners:
  - name: [username]@[domain]
    kind: User
```
everything worked fine and the user was able use the tenant.

Thanks again to @prometherion for taking the time to help with the issue!
